### PR TITLE
fix: prevent cache false hits for resources stuck in deletion

### PR DESCRIPTION
**Pull request by ugiordan**
## Description
Fixes cache bug in deploy action where objects with `deletionTimestamp` are cached normally instead of bypassing cache, causing false cache hits that prevent recreation of stuck Kubernetes resources.

**Root Cause**: The deploy action did not check for deletionTimestamp before performing cache lookups. When resources got stuck in deletion (due to finalizers), the deploy action would find cache hits for these terminating resources and skip redeployment, leaving them permanently stuck.

**Solution**: 
- Added `ShouldSkip` method to `Cache` struct that bypasses cache for objects with `deletionTimestamp`
- Added proactive cache cleanup via `Delete` method when objects have `deletionTimestamp`
- Refactored duplicated cache logic into `isCachedAndShouldSkip` helper method
- Made `newCache` function public (`NewCache`) for testing

**Link to JIRA**: https://issues.redhat.com/browse/RHOAIENG-34786

### Unit Tests
- `TestCacheCleanupVerification`: Verifies cache cleanup when `deletionTimestamp` is present
- `testDeletionTimestampBypassesCache`: Confirms cache bypass for objects with `deletionTimestamp`
- All existing cache tests pass

### Manual Testing on Live Cluster
- **KServe Component**: Reproduced cache bug using finalizers, confirmed UID persistence indicating deployment not recreated
- **ModelRegistry Component**: Verified same cache behavior affects multiple components
- **Fix Verification**: After applying fix, deployment UID changes correctly when finalizer removed, proving recreation works

### Reproduction Steps Verified
1. Enable component → record deployment UID
2. Add finalizer to deployment 
3. Disable component → deployment gets `deletionTimestamp` but stuck due to finalizer
4. Re-enable component → **Bug**: Same UID persists (not recreated)
5. **Fix**: Cache bypassed, proper recreation after finalizer removal

### Integration Tests
- Ran existing integration test pipeline - all tests pass
- New unit tests prevent regression of this cache behavior

## Screenshot or short clip
N/A - Cache behavior verified through UID tracking and unit tests.

## Merge criteria
- [x] You have read the [contributors guide](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/CONTRIBUTING.md).
- [x] Commit messages are meaningful - have a clear and concise summary and detailed explanation of what was changed and why.
- [x] Pull Request contains a description of the solution, a link to the JIRA issue, and to any dependent or related Pull Request.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work
- [x] The developer has run the integration test pipeline and verified that it passed successfully

### E2E test suite update requirement

When bringing new changes to the operator code, such changes are by default required to be accompanied by extending and/or updating the E2E test suite accordingly.

To opt-out of this requirement:
1. **Please inspect the [opt-out guidelines](https://github.com/opendatahub-io/opendatahub-operator/blob/main/docs/e2e-update-requirement-guidelines.md)**, to determine if the nature of the PR changes allows for skipping this requirement
2. If opt-out is applicable, provide justification in the dedicated `E2E update requirement opt-out justification` section below
3. Check the checkbox below:
- [x] Skip requirement to update E2E test suite for this PR
4. Submit/save these changes to the PR description. This will automatically trigger the check.

#### E2E update requirement opt-out justification
This PR fixes a cache bug in the deploy action that was causing existing E2E tests to fail intermittently. The fix:

1. **Internal implementation fix**: Changes are isolated to the caching mechanism and don't affect external APIs or behavior
2. **Improves existing E2E reliability**: Rather than requiring new E2E tests, this fix makes existing E2E tests more reliable by preventing resource recreation failures
3. **Comprehensive unit test coverage**: Added specific unit tests (`TestCacheCleanupVerification`, `testDeletionTimestampBypassesCache`) that thoroughly test the edge case
4. **Manual verification completed**: Extensive manual testing on live cluster with multiple components (KServe, ModelRegistry) confirms fix works
5. **No new functionality**: This is a bug fix that restores intended behavior, not new feature development

The fix addresses systematic E2E test failures, making the existing test suite more stable rather than requiring new E2E scenarios.

## Summary by CodeRabbit

* **Improvements**
  * Unified pre-deploy skip gating for consistent deployment decisions across resource types.
  * Simplified and standardized cache initialization for more reliable behavior.
* **Performance**
  * Fewer unnecessary deployments by reliably skipping unchanged or terminating resources.
* **Bug Fixes**
  * Resources marked for deletion are skipped and their cache entries are cleaned to prevent stale state.
* **Tests**
  * Added table-driven tests and helpers covering deletion-timestamp handling, cache cleanup, and TTL behavior.

### CODE REVIEW

File: pkg/controller/actions/deploy/action_deploy_cache_test.go [hunk 1]
```diff
@@ -263,3 +317,146 @@ func testCacheTTL(t *testing.T, cli client.Client, obj client.Object) {
...
+	current.SetDeletionTimestamp(&now)
+
+	// Update reconciliation request to use the object with deletionTimestamp
+	rr.Resources[0] = *current
+
+	// Test the fix: objects with deletionTimestamp should bypass cache
+	// Expected behavior: cache is bypassed and deployment proceeds
+	// In our test environment, this will result in a validation error because
+	// deletionTimestamp is immutable, but this proves cache bypass worked
+	err = action(ctx, &rr)
```
Comment (davidebianchi) [inline pkg/controller/actions/deploy/action_deploy_cache_test.go]: If the resource is in terminating state, the action try to deploy and fails because it is in terminating state?
There is some resource cleanup to resolve the terminating state before recreating the resource?

File: pkg/controller/actions/deploy/action_deploy_cache.go [hunk 2]
```diff
@@ -94,6 +94,40 @@ func (r *Cache) Has(original *unstructured.Unstructured, modified *unstructured.
...
+		return nil
+	}
+
+	return r.s.Delete(key)
+}
+
+// ShouldSkip determines whether resource deployment should be skipped based on cache state.
+// Returns true if the resource is cached and deployment should be skipped, false if deployment should proceed.
+// Always proceeds (bypasses cache) for objects with deletionTimestamp to ensure proper recreation of stuck resources.
+func (r *Cache) ShouldSkip(original *unstructured.Unstructured, modified *unstructured.Unstructured) (bool, error) {
```
Comment (lburgazzoli) [inline pkg/controller/actions/deploy/action_deploy_cache.go]: nit: I think we should find a better name as usually `Should` indicates a side effect free method, whereas this does have some side effect since it alters the cache

File: pkg/controller/actions/deploy/action_deploy.go [hunk 3]
```diff
@@ -170,6 +170,19 @@ func (a *Action) run(ctx context.Context, rr *odhTypes.ReconciliationRequest) er
 	return nil
 }
 
+// isCachedAndShouldSkip determines whether resource deployment should be skipped based on cache state.
+// Returns true if the resource is cached and deployment should be skipped, false if deployment should proceed.
+// Delegates to cache for deletion timestamp handling and cache cleanup.
+func (a *Action) isCachedAndShouldSkip(current *unstructured.Unstructured, desired *unstructured.Unstructured) (bool, error) {
```
Comment (lburgazzoli) [inline pkg/controller/actions/deploy/action_deploy.go]: nit: maybe just name it `ShouldSkip` as the `isCached` does not matter much

Comment (ugiordan) [reply pkg/controller/actions/deploy/action_deploy.go]: Sounds good.

File: pkg/controller/actions/deploy/action_deploy_cache.go (hunk 2 above)
Comment (ugiordan) [reply pkg/controller/actions/deploy/action_deploy_cache.go]: What about `CheckAndCleanup` or `ProcessCacheEntry`?

Comment (lburgazzoli) [reply pkg/controller/actions/deploy/action_deploy_cache.go]: No strong opinions, we can even leave it to a future PR

Comment (ugiordan) [reply pkg/controller/actions/deploy/action_deploy_cache.go]: Ok, for now I'll go with `ProcessCacheEntry`.

File: pkg/controller/actions/deploy/action_deploy_cache_test.go (hunk 1 above)
Comment (ugiordan) [reply pkg/controller/actions/deploy/action_deploy_cache_test.go]: > If the resource is in terminating state, the action try to deploy and fails because it is in terminating state?

Yes, and that's correct behavior. The cache fix ensures we attempt deployment on terminating objects (instead of incorrectly skipping them). The deployment fails as expected since you can't patch terminating objects.

> There is some resource cleanup to resolve the terminating state before recreating the resource?

No, and we shouldn't add any! Our operator just needs to:

1. Try to deploy (fails on terminating objects)
2. Retry later
3. Succeed once the object is fully deleted

Note: There's a separate issue where the DSC incorrectly reports `Ready` even when components are stuck in a terminating state. The DSC should report `NotReady` to reflect the actual state.

Comment (davidebianchi) [reply pkg/controller/actions/deploy/action_deploy_cache_test.go]: Ok great, thank you! I mainly wanted to be sure I understood the behaviour 😄  All clear now!

Comment (lburgazzoli) [reply pkg/controller/actions/deploy/action_deploy_cache_test.go]: Wonder if we should even avoid trying to deploy the object, the action should be re-triggered once the object gets deleted

Comment (ugiordan) [reply pkg/controller/actions/deploy/action_deploy_cache_test.go]: Good call. We should skip the deployment of terminating objects, but we can't do it silently because users need visibility. The DSC would show `NotReady` with a clear reason instead of hiding the issue.

Comment (lburgazzoli) [reply pkg/controller/actions/deploy/action_deploy_cache_test.go]: but this should not be the responsibility of the deploy action

Comment (ugiordan) [reply pkg/controller/actions/deploy/action_deploy_cache_test.go]: Totally agree. We can implement the skip logic, but I'm wondering how we can distinguish a correct deletion from a stuck one?

Comment (lburgazzoli) [reply pkg/controller/actions/deploy/action_deploy_cache_test.go]: I'm not sure we should.

We should certainly reflect the actual status but trying to find the actual reason my not be straightforward as there are many reasons for a resource not being removed. 

Maybe we can have the ComponentReady condition reporting status=False, reason=CleanupInProgress as long as the related internal API is present.

Comment (ugiordan) [reply pkg/controller/actions/deploy/action_deploy_cache_test.go]: That makes perfect sense! I can implement this in a separate PR and attach it to the DSC status JIRA.

Comment (ugiordan) [reply pkg/controller/actions/deploy/action_deploy_cache_test.go]: > Wonder if we should even avoid trying to deploy the object, the action should be re-triggered once the object gets deleted

Changed the logic to skip deployment entirely for objects with `deletionTimestamp` instead of attempting deployment and failing.
What changed:

- Deploy action now checks `deletionTimestamp` first and skips deployment
- Cache cleanup still happens to prevent stale entries
- Controller will retry naturally once the object is fully deleted

This avoids unnecessary deployment attempts and validation errors while keeping the cache clean.

File: pkg/controller/actions/deploy/action_deploy.go [hunk 4]
```diff
@@ -170,6 +170,36 @@ func (a *Action) run(ctx context.Context, rr *odhTypes.ReconciliationRequest) er
...
+	if desired == nil {
+		return false, nil
+	}
+
+	// Skip deployment if current object is terminating (Luca's suggestion)
+	// The action will be re-triggered once the object gets deleted
+	if current != nil && !current.GetDeletionTimestamp().IsZero() {
+		// Clean up cache if configured
+		if a.cache != nil {
+			if err := a.cache.CleanupIfTerminating(current, desired); err != nil {
```
Comment (lburgazzoli) [inline pkg/controller/actions/deploy/action_deploy.go]: do we actually need `CleanupIfTerminated` ? it seems it is doing the same check done here, so maybe just call delete ?

Comment (ugiordan) [reply pkg/controller/actions/deploy/action_deploy.go]: Make sense.

File: pkg/controller/actions/deploy/action_deploy_cache.go [hunk 5]
```diff
@@ -94,6 +94,28 @@ func (r *Cache) Has(original *unstructured.Unstructured, modified *unstructured.
...
+	if err != nil {
+		return fmt.Errorf("failed to compute cacheKey for deletion: %w", err)
+	}
+
+	if key == "" {
+		return nil
+	}
+
+	// Check if key exists before deleting to make Delete idempotent
+	_, exists, _ := r.s.GetByKey(key)
```
Comment (lburgazzoli) [inline pkg/controller/actions/deploy/action_deploy_cache.go]: is this actually required ? what would happen if you call Delete for an uncached object ?

File: pkg/controller/actions/deploy/action_deploy_cache_test.go [hunk 6]
```diff
@@ -67,55 +67,90 @@ func TestDeployWithCacheAction(t *testing.T) {
...
+			run: func(t *testing.T) {
+				t.Helper()
+				testCacheTTL(t, cli, createConfigMap())
+			},
+		},
+		{
+			name: "DeletionTimestampSkipsDeploymentAndCleansCache",
+			run: func(t *testing.T) {
+				t.Helper()
+				testDeletionTimestampHandling(t, cli, createDeployment())
```
Comment (lburgazzoli) [inline pkg/controller/actions/deploy/action_deploy_cache_test.go]: nit: deletion timestamp handling works for any type so maybe `createConfigMap`  should be enough

File: pkg/controller/actions/deploy/action_deploy_cache.go (hunk 5 above)
Comment (ugiordan) [reply pkg/controller/actions/deploy/action_deploy_cache.go]: Nope. It should already be idempotent. I’ll remove it.

File: pkg/controller/actions/deploy/action_deploy_cache.go [hunk 7]
```diff
@@ -94,6 +94,28 @@ func (r *Cache) Has(original *unstructured.Unstructured, modified *unstructured.
...
+	if !exists {
+		return nil // Already deleted or never existed - success
+	}
+	return r.s.Delete(key)
+}
```
[the remark this answers is not in the document]
Comment (ugiordan) [reply pkg/controller/actions/deploy/action_deploy_cache.go:112]: Nope. The reaourceVersion field is removed before computing the hash.

**Comment by ugiordan** [comment]
Overriding the e2e, not sure why the job restarts after the approval. 
/override ci/prow/opendatahub-operator-e2e
